app/services/katello/candlepin/consumer.rb

```diff
@@ -100,9 +100,13 @@ def pool_ids
       end
 
       def virtual_guests
+        return @virtual_guests unless @virtual_guests.nil?
         return [] if self.uuid.nil?
-        guest_uuids = Resources::Candlepin::Consumer.virtual_guests(self.uuid).map { |guest| guest['uuid'] }
-        ::Host.joins(:subscription_facet).where("#{Katello::Host::SubscriptionFacet.table_name}.uuid" => guest_uuids)
+        guest_ids = Resources::Candlepin::Consumer.virtual_guest_ids(self.uuid).map { |guest| guest['guestId'].downcase }
+        fact_name_id = RhsmFactName.find_by(:name => 'virt::uuid').id
+        @virtual_guests = ::Host.joins(:fact_values).
+                                 where("lower(#{FactValue.table_name}.value) IN (?) AND #{FactValue.table_name}.fact_name_id = ?", guest_ids, fact_name_id).
```

Comment:
Hmm... I am kind of worry this query when the facts table are extremely large.

Reply:
I agree with you. Maybe it would be better to go back to your original implementation and remove the virtual_guest_ids call.

I think there are two possible future improvements:

1) store the guest id on SubscriptionFacet for fast lookup based on /guestids API
2) new Candlepin API to retrieve guest UUIDs only (this is my preference)

Reply:
I filed a bug for Candlepin to provide an API for us https://bugzilla.redhat.com/show_bug.cgi?id=1772159

In the future we can switch to use it

Reply:
Great! Thanks for creating the bug.  I have rewinded the commit. Let me know if you found any issue.